### ISSUE TITLE
Error in ZMQ when running many repeated acquisitions

### ISSUE THREAD

**Issue by dpshepherd** [micro-manager/pycro-manager#286]
### Bug report

**Bug summary**

If you run many (in this case it occurred after 199) repeated acquisitions with a `hook_fn` (`post_camera_hook_fn` in this case), the following error comes up
```
Exception in thread AcquisitionHook:
Traceback (most recent call last):
  File "C:\Users\dpsheph1\Anaconda3\envs\fixed_pycro\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "C:\Users\dpsheph1\Anaconda3\envs\fixed_pycro\lib\threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\dpsheph1\Anaconda3\envs\fixed_pycro\lib\site-packages\pycromanager\acquire.py", line 145, in _acq_hook_startup_fn
    push_socket = bridge._connect_push(pull_port)
  File "C:\Users\dpsheph1\Anaconda3\envs\fixed_pycro\lib\site-packages\pycromanager\core.py", line 320, in _connect_push
    return DataSocket(
  File "C:\Users\dpsheph1\Anaconda3\envs\fixed_pycro\lib\site-packages\pycromanager\core.py", line 30, in __init__
    self._socket.bind("tcp://{}:{}".format(ip_address, port))
  File "C:\Users\dpsheph1\Anaconda3\envs\fixed_pycro\lib\site-packages\zmq\sugar\socket.py", line 172, in bind
    super().bind(addr)
  File "zmq\backend\cython\socket.pyx", line 540, in zmq.backend.cython.socket.Socket.bind
  File "zmq\backend\cython\checkrc.pxd", line 28, in zmq.backend.cython.checkrc._check_rc
zmq.error.ZMQError: Permission denied
```

This is new, as we have run this many acquisitions before without an issue. We are running other hardware in-between acquisitions that do not easily fit into the `hook_fn` specification. For this particular microscope, it is a homemade fluidics exchange setup.

**Code for reproduction**

```
import numpy as np 
from pycromanager import Acquisition

def camera_hook_fn(event,bridge,event_queue):

    core = bridge.get_core()

    # do something with core

    return event

events = []
for index, z_um in enumerate(np.arange(start=0, stop=10, step=0.5)):
    evt = {
            #'axes' is required. It is used by the image viewer and data storage to
            #identify the acquired image
            'axes': {'z': index},

            #the 'z' field provides the z position in µm
            'z': z_um}
    events.append(evt)

N = 200 # increase as needed to generate error

for i in range(N):

  # python code to run other hardware
  with Acquisition('/path/to/storage', 'testacq', post_camera_hook_fn=camera_hook_fn) as acq:
      acq.acquire(events)
  acq = None
```

**Expected outcome**

N Acquisitions successfully complete.

**Actual outcome**

N Acquisitions do not successfully complete.

**Version Info**

  * Operating system: Windows 10 64-bit
  * pycromanager version: 0.12.8
  * MicroManager version: Nightly build 20210510
  * Python version: 3.8
  * Python environment (command line, IDE, Jupyter notebook, etc): VS Code and command line

**Comment by dpshepherd** [comment]
This is different than the error message that always comes up on Windows when using a hook_fn, but does not cause the acquisition to hang.

```
Exception ignored in: <function JavaObjectShadow.__del__ at 0x000001F2DD222430>
Traceback (most recent call last):
  File "C:\Users\dpsheph1\Anaconda3\envs\fixed_pycro\lib\site-packages\pycromanager\core.py", line 470, in __del__
    self._close()
  File "C:\Users\dpsheph1\Anaconda3\envs\fixed_pycro\lib\site-packages\pycromanager\core.py", line 460, in _close
    self._socket.send(message)
  File "C:\Users\dpsheph1\Anaconda3\envs\fixed_pycro\lib\site-packages\pycromanager\core.py", line 115, in send
    self._socket.send_multipart(message_parts)
  File "C:\Users\dpsheph1\Anaconda3\envs\fixed_pycro\lib\site-packages\zmq\sugar\socket.py", line 555, in send_multipart
    return self.send(msg_parts[-1], flags, copy=copy, track=track)
  File "C:\Users\dpsheph1\Anaconda3\envs\fixed_pycro\lib\site-packages\zmq\sugar\socket.py", line 505, in send
    return super(Socket, self).send(data, flags=flags, copy=copy, track=track)
  File "zmq\backend\cython\socket.pyx", line 718, in zmq.backend.cython.socket.Socket.send
  File "zmq\backend\cython\socket.pyx", line 759, in zmq.backend.cython.socket.Socket.send
  File "zmq\backend\cython\socket.pyx", line 135, in zmq.backend.cython.socket._check_closed
zmq.error.ZMQError: Unknown error
```

**Comment by henrypinkard** [comment]
I suspect its something to do the the ZMQ sockets not cleaning up fast enough in between acquisitions. How long does your `# python code to run other hardware` take to execute? If its short, does introducing longer delays fix the problem or increase the `N` that it takes to reliably trigger it?

Also, is there a reason you couldnt just run this all as one big acquisition?

Instead of:

```python
for i in range(N):
  # python code to run other hardware
  with Acquisition('/path/to/storage', 'testacq', post_camera_hook_fn=camera_hook_fn) as acq:
      acq.acquire(events)
  acq = None
```

Could you do?

```python
with Acquisition('/path/to/storage', 'testacq', post_camera_hook_fn=camera_hook_fn) as acq:
    for i in range(N):
        #make new 'events' that vary along some axis here
        acq.acquire(events)
       # python code to run other hardware
```

Maybe you'd want to ensure that your events had finished executing before adding in a new batch. You could do this with the newly added `storage_monitor_callback_fn`, which I havent put in docs yet (but [its in the API](https://pycro-manager.readthedocs.io/en/latest/apis.html#pycromanager.Acquisition))

**Comment by dpshepherd** [comment]
> I suspect its something to do the the ZMQ sockets not cleaning up fast enough in between acquisitions and then zmq trying to reuse ports that have been designated for other purpose (based on [this](https://stackoverflow.com/questions/29950824/pyzmq-permission-denied)).

Ok, that makes sense.

> 
> How long does your `# python code to run other hardware` take to execute? If its short, does introducing longer delays fix the problem or increase the `N` that it takes to reliably trigger it?

There are two levels of timing between repeated acquisition calls.
1. The faster timing is the time is takes the scanning stage to reset and change the hardware configuration, which is seconds. This occurs between 10-50 times in rapid succession. 
2. The slow timing is the time it takes the fluidics unit to run, which varies from 10 minutes to 60 minutes depending on the experiment. Here, the microscope sits while the fluidics are running and a closed-loop controller outside of Python actively maintains the stage position.

We have taken over setting up the hardware for each "YZC" acquisition from the `Acquisition `engine, because we want an individual directory for each "YZC" acquisition.  We move the stage in the X direction during our acquisition, but we setup that up using hardware triggering and start it using the `post_camera_hook_fn` .The main reason for this is that call to an `Acquisition` generates >100 GB. We still can't load these large acquisitions efficiently using the Dask feature of the `Dataset` object. Breaking up the acquisitions allows to us move each "YZC" over the network to a server after it completes and still load via Dask.

There are some other practical reasons we like controlling the "YZC" setup outside of "Acquisition" because of the constant speed stage scan, hardware triggering, etc... that we have discussed before. Our microscope definition doesn't really match the usual definition used within Micro-manager / Pycro-manager for Acquisitions since it is a single objective light sheet that we are stage scanning for millions of images.

> 
> Also, is there a reason you couldnt just run this all as one big acquisition?

It's possible, see explanation above about why we are hesitant to do this. If the `Dataset` loading issue can be solved, then this would be less of a problem. Since that appears to be an issue with Dask that is out of all of our control, we are a bit stuck there.

> 
> Instead of:
> 
> ```python
> for i in range(N):
>   # python code to run other hardware
>   with Acquisition('/path/to/storage', 'testacq', post_camera_hook_fn=camera_hook_fn) as acq:
>       acq.acquire(events)
>   acq = None
> ```
> 
> Could you do?
> 
> ```python
> with Acquisition('/path/to/storage', 'testacq', post_camera_hook_fn=camera_hook_fn) as acq:
>     for i in range(N):
>         #make new 'events' that vary along some axis here
>         acq.acquire(events)
>        # python code to run other hardware
> ```
> 
> Maybe you'd want to ensure that your events had finished executing before adding in a new batch. You could do this with the newly added `storage_monitor_callback_fn`, which I havent put in docs yet (but [its in the API](https://pycro-manager.readthedocs.io/en/latest/apis.html#pycromanager.Acquisition))

We'll take a look and see if we can modify it this way and still load the data efficiently.

**Comment by henrypinkard** [comment]
You can load `Dataset`s that slice along axes now (new feature). So something like `dataset.as_array(z=5, y=7)` should allow you to open data one RAM-sized piece at a time.

If you ever want to look into making a minimal example using dask only that stacks together multiple memory-mapped chunks using other dask functions than the one `dataset.as_array` uses and that doesnt overload RAM, I'd be happy to implement it into the `Dataset class`. Don't have the bandwidth to dig into it myself right now though

**Comment by dpshepherd** [comment]
> You can load `Dataset`s that slice along axes now (new feature). So something like `dataset.as_array(z=5, y=7)` should allow you to open data one RAM-sized piece at a time.

Ok, we'll check it out. The issue is not RAM (servers have 1 or 2 TB respectively). The issue is it takes an extremely long time to load a dataset as the number of images increases. The time required for large datsets is crippling for downstream processing and analysis of the data if we don't keep each dataset relatively small.

Turns out one of the lab members previously implemented some helper functions to read all needed images using `dataset.read_image()` and let me know this is orders of magnitude faster than using Dask. We will explore this alternative and see if it works.

> If you ever want to look into making a minimal example using dask only that stacks together multiple memory-mapped chunks using other dask functions than the one `dataset.as_array` uses and that doesnt overload RAM, I'd be happy to implement it into the `Dataset class`. Don't have the bandwidth to dig into it myself right now though

We are slammed as well, so haven't had time to dive into the memory-mapped chunks either. It's on the long list of things to look into it.

Thanks!

**Comment by dpshepherd** [comment]
Is there a usage example of `storage_monitor_callback_fn`? It isn't entirely clear to me how use this function to monitor if the events had finished executing from the API.

Could we just use `acq.await_completion()` and do away with the `with `block?

**Comment by henrypinkard** [comment]
> Ok, we'll check it out. The issue is not RAM (servers have 1 or 2 TB respectively). The issue is it takes an extremely long time to load a dataset as the number of images increases. The time required to process a large acquisition is crippling for downstream processing and analysis of the data if we don't keep each individual dataset relatively small.
> 
> Turns out one of the lab members previously implemented some helper functions to read all needed images using `dataset.read_image()` and let me know this is orders of magnitude faster than using Dask. We will explore this alternative and see if it works.

If dasks problem is a suprlinear scaling with number of chunks, then slicing with the new feature would help. Otherwise if theres just a constant per-chunk overhead, then `read_image` will be the way to go I think. Easy enough to test this. I don't know enough about dask internals to know if this would change



> Is there a usage example of `storage_monitor_callback_fn`? It isn't entirely clear to me how use this function to monitor if the events had finished executing from the API.
> 
> Could we just use `acq.await_completion()` and do away with the `with `block?

`acq.await_completion()` will only return after all events are done. So if you're doing one big acquisition and sending in events multiple times, this won't give you a callback for intermediate events finishing.

Use `storage_monitor_callback_fn` like this (I think, untested):

```python
def storage_monitor_callback_fn(axes):
   #figure out when image from the last event you added is saved
   if axes['z'] == last_z_index and axes['y'] == last_y_index:
      # do whatever you need to do in between event batches
      acq.acquire(next_batch_of_events)

with Acquisition(storage_monitor_callback_fn=storage_monitor_callback_fn) as acq:
  acq.acquire(first_batch of events)


```

**Comment by dpshepherd** [comment]
> > Is there a usage example of `storage_monitor_callback_fn`? It isn't entirely clear to me how use this function to monitor if the events had finished executing from the API.
> > Could we just use `acq.await_completion()` and do away with the `with `block?
> 
> `acq.await_completion()` will only return after all events are done. So if you're doing one big acquisition and sending in events multiple times, this won't give you a callback for intermediate events finishing.

Perhaps a dumb question. If we do call `acq.await_completion()`, does this enforce that the ZMQ socket cleans up after a call to an `Acquisition`? Or does the ZMQ socket clean up happen separately?

**Comment by henrypinkard** [comment]
Im pretty sure that won't happen until the python garbage collector runs on the `Acquisition` object, which I believe can only happen after `acq = None`. Alternatively, you can call `del acq`, which maybe would change how fast this process happens. It's a little unclear to me based on the python docs.

If you can't get anything to work, I could potentially change the behavior for allocating new sockets to reuse ports less often. Obviously you couldn't do this indefinitely or you'd run out of numbers, but it might seem reasonable to cycle through a range of say 20 port numbers before going back to the original. This would give it a lot more time to clean up

**Comment by dpshepherd** [comment]
Got it.

We call `acq = None` right now in the real and demo code that produces the error, so I don't think that will make a difference.  We can also try `del acq`.

I think we should be able to figure this out with the existing library, we just need to think a bit more carefully and do some testing on how we are setting things up.

Thanks!

**Comment by dpshepherd** [comment]
Finally had some time to test `storage_monitor_callback_fn`.

We are having trouble getting the acquisition to return and wrote a simpler script which just checks if the acquisition is done writing using the demo config. This is using MM 2.0 Nightly Build 20210602 and Pycro-manager 0.12.9.

```python
#!/usr/bin/env python

'''
Test storage_monitor_callback_fn()

Shepherd 06/21
'''

# imports
from pycromanager import Bridge, Acquisition
from pathlib import Path
from time import sleep
from functools import partial
import gc

# this function is executed once the acquisition engine has setup
def post_hook_fn(event,bridge,event_queue):

    # open bridge to MM
    bridge = Bridge()
    core = bridge.get_core()

    return event

# this function makes sure that all data is written to disk
def storage_monitor_callback_fn(final_z,final_c,final_e,axes):
    
    global data_storage_finished

    print(axes['z'],axes['c'],axes['e'])

    if axes['z']==(final_z-1) and axes['c']==(final_c-1) and axes['e']==(final_e-1):
        print('Setting flag, writing last image.')
        data_storage_finished = True


def main():

    # global flag that data storage is finished
    global data_storage_finished

    # create links to MM
    bridge = Bridge()
    core = bridge.get_core()

    events = []
    num_z=10
    num_e=3
    num_c=4
    for z in range(num_z):
        for etl in range(num_e):
            for c in range(num_c):        
                evt = { 'axes': {'z': z, 'e': etl, 'c': c}}
                events.append(evt)

    # construct partial functions for acquisition hooks
    storage_monitor_callback_metadata_fn = partial(storage_monitor_callback_fn, num_z, num_c, num_e)

    data_storage_finished = False
        
    save_name = Path('test')
    save_directory = Path('c:\\storage_monitor_test\\')
    with Acquisition(directory=str(save_directory), name=str(save_name), post_hardware_hook_fn=post_hook_fn, 
                    show_display=False, max_multi_res_index=0, storage_monitor_callback_fn=storage_monitor_callback_metadata_fn) as acq:
        acq.acquire(events)

    # wait until all data is written to disk
    while not(data_storage_finished):
        sleep(1)
        print('Storage not finished')

    # clean up this acquisition object and force Python to run garbage collector
    acq = None
    gc.collect()

# run
if __name__ == "__main__":
    main()
```

This gives the following output:

```python
0 0 0
0 1 0
0 2 0
0 3 0
0 0 1
0 1 1
0 2 1
0 3 1
0 0 2
0 1 2
0 2 2
0 3 2
1 0 0
1 1 0
1 2 0
1 3 0
1 0 1
1 1 1
1 2 1
1 3 1
1 0 2
1 1 2
1 2 2
1 3 2
2 0 0
2 1 0
2 2 0
2 3 0
2 0 1
2 1 1
2 2 1
2 3 1
2 0 2
2 1 2
2 2 2
2 3 2
3 0 0
3 1 0
3 2 0
3 3 0
3 0 1
3 1 1
3 2 1
3 3 1
3 0 2
3 1 2
3 2 2
3 3 2
4 0 0
4 1 0
4 2 0
4 3 0
4 0 1
4 1 1
4 2 1
4 3 1
4 0 2
4 1 2
4 2 2
4 3 2
5 0 0
5 1 0
5 2 0
5 3 0
5 0 1
5 1 1
5 2 1
5 3 1
5 0 2
5 1 2
5 2 2
5 3 2
6 0 0
6 1 0
6 2 0
6 3 0
6 0 1
6 1 1
6 2 1
6 3 1
6 0 2
6 1 2
6 2 2
6 3 2
7 0 0
7 1 0
7 2 0
7 3 0
7 0 1
7 1 1
7 2 1
7 3 1
7 0 2
7 1 2
7 2 2
7 3 2
8 0 0
8 1 0
8 2 0
8 3 0
8 0 1
8 1 1
8 2 1
8 3 1
8 0 2
8 1 2
8 2 2
8 3 2
9 0 0
9 1 0
9 2 0
9 3 0
9 0 1
9 1 1
9 2 1
9 3 1
9 0 2
9 1 2
9 2 2
9 3 2
Setting flag, writing last image.
```

And never finishes the acquisition. Is there something we need to do in `storage_monitor_callback_fn`. to tell the acquisition to finish up?

Thanks!

**Comment by dpshepherd** [comment]
Also, the NDTiIff and Index files on the disk are not finished up when I run the above code.

**Comment by henrypinkard** [comment]
I took a fresh look at this and found:

1) You're right, there's definitely a bug right now with `storage_monitor_callback_fn`, which is that the files cant be properly closed and have their lengths set because they are open in another process. I think this is only a windows bug (did you test in on your linux machine too?) I finally have a windows computer to develop on, so hopefully I can figure it out and there will be fewer things like this in the future.

2) Even without using the storage monitoring, it would definitely be more intuitive that exiting with `with` block means everything is cleaned up and a new acquisition can be started. I wasn't able to actually reproduce this on my own, but I [added some more checks](https://github.com/micro-manager/pycro-manager/pull/291) for `Acquisition` to wait on the other threads it spawns before exiting the block, which I think might be the source of this problem. Changes will be in 0.12.10. Could you give it a try?

**Comment by henrypinkard** [comment]
>     1. You're right, there's definitely a bug right now with `storage_monitor_callback_fn`, which is that the files cant be properly closed and have their lengths set because they are open in another process. I think this is only a windows bug (did you test in on your linux machine too?) I finally have a windows computer to develop on, so hopefully I can figure it out and there will be fewer things like this in the future.

This should be fixed in https://github.com/micro-manager/pycro-manager/pull/292 and available in new pip version.

Let me know when you have a chance to test the previous test to see if it fixed the issue

**Comment by dpshepherd** [comment]
The `with Acquisition` block now takes a lot longer to exit. For highs-speed acquisitions the disk activity stops, there do not appear to be any images in memory, and then there is a minutes delay until the `with` block returns. We no longer get the ZMQ related warnings/errors we were previously getting. However, apparently we occasionally get  a new type of ZMQ error that renders MM unresponsive to pycro-manager. I wasn't in the lab when they got the new error, so I need replicate on my own / have them log the error.

Haven't had a chance to try `storage_monitor_callback_fn `again yet.

**Comment by henrypinkard** [comment]
Hmmm sounds like the ZMQ shutdown might take a long time for some reason. The other you might try, is to manually control the port the acquisition uses. This is a feature I just added. So if the theory that collisions on the same port number are causing the failures of successive acquisitions is true, you should be able to manually cycle through different ports to give more time in between for things to finish up. 

E.g. 
`Acquisition(port=4827)`
`Acquisition(port=4837)`
`Acquisition(port=4847)`
`Acquisition(port=4827)`

You need to give some space in between numbers because acq events, hooks, processors, etc each use their own ports, counting up successively from that starting one. So `Acquisition(port=4827)` might put events on 4828, hook on 4829, etc.

When you have a chance to test you can also playing around with commenting out lines 534-544 https://github.com/micro-manager/pycro-manager/blob/390e435f5f111f03dc5d8de4ba9ec81e20e5bf9c/pycromanager/acquire.py#L534, which where the post acquisition delays to wait on ZMQ are likely coming from

**Comment by dpshepherd** [comment]
Did you test with using hook functions on windows? When we use `post_camera_hook_fn`, we now get:

```
Exception ignored in: <function JavaObjectShadow.__del__ at 0x000002224670C280>
Traceback (most recent call last):
  File "C:\Users\qi2lab\miniconda3\envs\OPM\lib\site-packages\pycromanager\core.py", line 474, in __del__
    self._close()
  File "C:\Users\qi2lab\miniconda3\envs\OPM\lib\site-packages\pycromanager\core.py", line 464, in _close
    self._socket.send(message)
  File "C:\Users\qi2lab\miniconda3\envs\OPM\lib\site-packages\pycromanager\core.py", line 115, in send
    self._socket.send_multipart(message_parts)
  File "C:\Users\qi2lab\miniconda3\envs\OPM\lib\site-packages\zmq\sugar\socket.py", line 555, in send_multipart
    return self.send(msg_parts[-1], flags, copy=copy, track=track)
  File "C:\Users\qi2lab\miniconda3\envs\OPM\lib\site-packages\zmq\sugar\socket.py", line 505, in send
    return super(Socket, self).send(data, flags=flags, copy=copy, track=track)
  File "zmq\backend\cython\socket.pyx", line 718, in zmq.backend.cython.socket.Socket.send
  File "zmq\backend\cython\socket.pyx", line 759, in zmq.backend.cython.socket.Socket.send
  File "zmq\backend\cython\socket.pyx", line 135, in zmq.backend.cython.socket._check_closed
zmq.error.ZMQError: not a socket
```

Newest pycro-manager and newest nightly build. Windows 10 64-bit.

**Comment by dpshepherd** [comment]
Now I'm wondering if this is related to issue #298? The instrument is tied up with a long run, but I can check if this behavior reappeared with the newest version and was gone before once it is done.

**Comment by dpshepherd** [comment]
Ok, this error is also present in version 0.12.10 when using a `post_camera_hook_fn` and Micromanager nightly build 20210628. I haven't gone back to the older MM builds yet where you hadn't yet pushed updates to the Java part.

**Comment by henrypinkard** [comment]
I think I finally tracked this down. There was a low level error with the shutdown of `Bridge`s. The cleanup behavior was erroneously in the destructor, which wasn't always called as I expected. I believe its fixed now by https://github.com/micro-manager/pycro-manager/pull/299. You now need to use `Bridge` in a `with` statement, or explicitly call `close()` when finished with it. Available in v0.13.0

Pretty sure this is the same issue as #298

I also wonder if this also fixes: https://github.com/micro-manager/pycro-manager/issues/169

**Comment by henrypinkard** [comment]
Thanks for tracking down how to reproduce this so well!

**Comment by dpshepherd** [comment]
We will give this a shot soon, thanks so much for the persistence on looking into this!

Two questions:

1. For experiments where we want to run repeated `with Acquisition()` blocks, do we need to create/destroy a `Bridge `object for each one? Or can we just use one `Bridge `object for all of them?
2. This means in the `post_camera_hook_fn`, we should just create the local `Bridge` object in the function using a `with `block and that will take care of create/destroy within the hook function itself? Do you anticipate any timing issues from this?

**Comment by henrypinkard** [comment]
>     1. For experiments where we want to run repeated `with Acquisition()` blocks, do we need to create/destroy a `Bridge `object for each one? Or can we just use one `Bridge `object for all of them?

Either one should work now thats its fixed, but probably easier to just stick with a single one.

>     2. This means in the `post_camera_hook_fn`, we should just create the local `Bridge` object in the function using a `with `block and that will take care of create/destroy within the hook function itself? Do you anticipate any timing issues from this?

You could create a new `Bridge` in each hook call, and I think it would clean up fine now, but you're correct in thinking that there is some overhead every time you create a new bridge so you actually probably don't want to do this. The 3 argument hook function signature (`hook_fn(event, bridge, event_queue)` instead of `hook_fn(event)`) should solve this though, because it keeps a single instance of the `Bridge` that's created at the start of acquisition and passed into the hook every time its called. It looks like from your code in the first post you've aready got this setup correctly

**Comment by dpshepherd** [comment]
Got it. I misunderstood and thought we needed to start declaring a new `Bridge` in the hook function now. My mistake.

**Comment by dpshepherd** [comment]
The ZMQ errors seem to be gone and `bridge.close()` worked as expected. We'll test more, but I'm closing this as it seems to be fixed. Thanks!!

**Comment by dpshepherd** [comment]
We just ran into a error that I think is related to our previous discussion on this issue. We are randomly seeing the following ZMQ error when running hundreds of repeated acquisitions:

```python
Exception in thread AcquisitionHook:
Traceback (most recent call last):
  File "C:\Users\qi2lab\miniconda3\envs\tilingSPIM-v2\lib\threading.py", line 954, in _bootstrap_inner
    self.run()
  File "C:\Users\qi2lab\miniconda3\envs\tilingSPIM-v2\lib\threading.py", line 892, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\qi2lab\miniconda3\envs\tilingSPIM-v2\lib\site-packages\pycromanager\acquire.py", line 151, in _acq_hook_startup_fn
    push_socket = bridge._connect_push(pull_port)
  File "C:\Users\qi2lab\miniconda3\envs\tilingSPIM-v2\lib\site-packages\pycromanager\core.py", line 340, in _connect_push
    return DataSocket(
  File "C:\Users\qi2lab\miniconda3\envs\tilingSPIM-v2\lib\site-packages\pycromanager\core.py", line 29, in __init__
    self._socket.bind("tcp://{}:{}".format(ip_address, port))
  File "C:\Users\qi2lab\miniconda3\envs\tilingSPIM-v2\lib\site-packages\zmq\sugar\socket.py", line 173, in bind
    super().bind(addr)
  File "zmq/backend/cython/socket.pyx", line 542, in zmq.backend.cython.socket.Socket.bind
  File "zmq/backend/cython/checkrc.pxd", line 28, in zmq.backend.cython.checkrc._check_rc
    raise ZMQError(errno)
zmq.error.ZMQError: Permission denied
```

We are using a post hardware hook and can reproduce using the demo config as follows (may need to increase number of times the acquisition is run to introduce crash):

```python
#!/usr/bin/env python

'''
Test ZMQerror with demo config

Shepherd 07/21
'''

# imports
from pycromanager import Bridge, Acquisition
from pathlib import Path
from time import sleep
import gc

# this function is executed once the acquisition engine has setup
def post_hook_fn(event,bridge,event_queue):

    # open bridge to MM
    core = bridge.get_core()
    time.sleep(1)
    return event

def main():

    # create links to MM
    bridge = Bridge()
    core = bridge.get_core()

    events = []
    num_z=205
    num_e=6
    num_c=1
    for acq_idx in range(num_z):
        for etl in range(num_e):
            for c in range(num_c):        
                evt = { 'axes': {'z': z, 'e': etl, 'c': c}}
                events.append(evt)

    for tile_idx in range(0,300):

        core.clear_circular_buffer() # our camera hardware needs this, so include in the demo config code as well
        save_name = Path('test_'+str(tile_idx).zfill(4))
        save_directory = Path('c:\\repeated_acq_test\\')
        with Acquisition(directory=str(save_directory), name=str(save_name), post_hardware_hook_fn=post_hook_fn, 
                        show_display=False, max_multi_res_index=0) as acq:
            acq.acquire(events)

        # clean up this acquisition object and force Python to run garbage collector
        acq = None
        gc.collect()

# run
if __name__ == "__main__":
    main()
```

pycromanager version: 22.1.0
MM 2.0 version: 20210712
Windows 10 64-bit
ImageJ has 32 gb of RAM and MM sequence buffer has 12 gb of RAM allocated

**Comment by henrypinkard** [comment]
I wasn't able to reproduce, but it seems like its definitely because of a race condition in closing the socket's hook.

Thought this was all fixed, but I added in another layer of checks to the new version on pip to make sure.

If you still see it, could you also run `Acquisition(debug=True)`, because it now prints the port number when it closes so, this shoudl be confirmed. But maybe try without that to start because it probably alters the timing

**Comment by dpshepherd** [comment]
We've been testing if the core object that is created in the hook function is not closed correctly and eventually accumulates, but not clear yet if it is that.

Should have time this week to test with `Acquisition(debug=True)`.

Thanks! I know this is a weird use case, but it's super helpful on this particular instrument.

**Comment by kidantad** [comment]
Hello there. I am having a similar issue when using the the post hook function. I had this error for sometime but it wasn't affecting my acquisition, but now it seems like I've missed my stage and camera synchronization. After reviewing this thread, I tried explicitly calling close() when finished with acquisition and/or adding " with Bridge as bridge():" in my hook but that generated its own error. I also added the cleaning up the acquisition object and garbage collector suggested by @dpshepherd but that didn't make any difference. I'm curious to know if you were able to solve the "ZMQError: not a scocket" issue associated with the post hook function.

**For reference my post camera hook code is as follow and is similar to my post hardware hook:**

def post_camera_hook_(event,bridge,event_queue):

 
    # acquire core from bridge
    mmc = bridge.get_core()
    
    #new
    return event

    # send MS2000 command to start constant speed scan
    command='SCAN'
    mmc.set_property('XYStage','SerialCommand',command)

    return event

ready = mmc.get_property('XYStage','SerialResponse')

**Code for final acquisition:**

for z in z_sequence:
    with Acquisition(directory=save_dir, name=save_name, post_hardware_hook_fn=post_hardware_hook,
                    post_camera_hook_fn=post_camera_hook_, show_display=True, max_multi_res_index=0) as acq:
        events = multi_d_acquisition_events(
                                        num_time_points=400, time_interval_s=0,
                                        z_start =z, z_end=z-.5*z_scan_step, z_step=z_scan_step,
        )
        
        
        acq.acquire(events)
        del acq
        gc.collect()
        
**And the ERROR I get is:**
Exception ignored in: <function JavaObjectShadow.__del__ at 0x000001A351BECE50>
Traceback (most recent call last):
  File "c:\users\nikon_1\appdata\local\programs\python\python38\lib\site-packages\pycromanager\core.py", line 457, in __del__
    self._close()
  File "c:\users\nikon_1\appdata\local\programs\python\python38\lib\site-packages\pycromanager\core.py", line 447, in _close
    self._socket.send(message)
  File "c:\users\nikon_1\appdata\local\programs\python\python38\lib\site-packages\pycromanager\core.py", line 102, in send
    self._socket.send_multipart(message_parts)
  File "c:\users\nikon_1\appdata\local\programs\python\python38\lib\site-packages\zmq\sugar\socket.py", line 463, in send_multipart
    return self.send(msg_parts[-1], flags, copy=copy, track=track)
  File "c:\users\nikon_1\appdata\local\programs\python\python38\lib\site-packages\zmq\sugar\socket.py", line 416, in send
    return super(Socket, self).send(data, flags=flags, copy=copy, track=track)
  File "zmq\backend\cython\socket.pyx", line 728, in zmq.backend.cython.socket.Socket.send
  File "zmq\backend\cython\socket.pyx", line 769, in zmq.backend.cython.socket.Socket.send
  File "zmq\backend\cython\socket.pyx", line 135, in zmq.backend.cython.socket._check_closed
zmq.error.ZMQError: not a socket

     
Thank you.

**Comment by henrypinkard** [comment]
This thread is a bit out of date, since in the newest version you don't need to explicitly create or call `Bridge`. So I would first make sure youre up to date on versions and their APIs.

Next, I would test if you're able to find the same behavior on the demo configuration, which would rule out a problem with your specific hardware.

**Comment by henrypinkard** [comment]
> from pycromanager import Core core = Core()
> 
> I get an error message saying that it isn't able to import core.

What is the error message?

With the latest version, you should never be creating a `Bridge` yourself.

**Comment by henrypinkard** [comment]
Im a but confused about what you mean and where your confusion lies. Can you please post your full code (maybe on a new thread) along with expected and actual behavior?

**Comment by kidantad** [comment]
I'm sorry about that. I have added my question in more detail and code after updating my system in issue #413 . Thank you in advance for your help.

**Comment by henrypinkard** [comment]
Pretty sure this is fixed now based on #428
